src/software/thunderscope/binary_context_managers/game_controller.py

```diff
@@ -27,17 +27,33 @@ class Gamecontroller(object):
     REFEREE_IP = "224.5.23.1"
     CI_MODE_OUTPUT_RECEIVE_BUFFER_SIZE = 9000
 
-    def __init__(self, supress_logs: bool = False) -> None:
+    def __init__(
+        self,
+        supress_logs: bool = False,
+        gamecontroller_port: int = None,
+        referee_addresse: int = None,
+    ) -> None:
         """Run Gamecontroller
 
         :param supress_logs: Whether to suppress the logs
+        :param gamecontroller_port: the port that the gamecontroller would bind to 
+        :param referee_address: the address the referee binds to
         """
         self.supress_logs = supress_logs
 
         # We need to find 2 free ports to use for the gamecontroller
         # so that we can run multiple gamecontroller instances in parallel
         self.referee_port = self.next_free_port()
-        self.ci_port = self.next_free_port()
+
+        self.REFEREE_IP = referee_addresse
+        if referee_addresse is None:
+            self.REFEREE_IP = "224.5.23.1"
+
```

Comment:
`REFEREE_IP` is supposed to be a constant, we shouldn't be modifying it inside here

Reply:
> On the other hand, https://github.com/UBC-Thunderbots/Software/issues/3160 discusses the problem we faced during Robocup where most teams connect their computer to the same network which has access to the internet. Furthermore, during a game or field testing, all teams on the field connect to the same LAN. Our goal is to basically almost always launch the gamecontroller on some non-official port/multicast IP address that other teams or other members of our team aren't listening to for gamecontroller commands. This will allow us to send gamecontroller commands that only effects the fullsystem being ran locally. This should be the case UNLESS we're playing an actual game where we're not launching our own gamecontroller locally and we're using the provided field gamecontroller. Hence https://github.com/UBC-Thunderbots/Software/issues/3161 discussing a new "robocup" mode used during actual games with a non-local gc. To achieve this, we will probably need to update RobotCommunication aswell to not always listen to the official gamecontroller/referee port/IP

doesn't this contradicts what nima said? am I misunderstanding something? are we setting the port instead then but keeping the mulitcast address constant?

Reply:
Nima's right but in the implementation, don't modify `REFEREE_IP`. That is a constant (which is why it's all capitalized) we use during real games for connecting to the Gamecontroller and we should leave it so that it always point to that ip address. By convention, we never modify constants during runtime.

Instead, use a new local variable:
```suggestion
        multicast_address = referee_address if referee_address else Gamecontroller.REFEREE_IP

```

Reply:
Shouldn't this be `self.multicast_address` instead of `multicast_address`, since we are accessing this variable in line 180?

Can I also change the name to `DEFAULT_REFEREE_IP` instead?

Reply:
Yeah, you're correct